File: getPopulation.m
function population=getPopulation(class, input,nPop)
[center,U] = fcm(input, 3, [2 100 1e-5]); %center = center cluster, U = membership level
[~, Hi] = max(U); % Yy = max value between both membership function, Hi = the class corresponding to the max value
[~, features] = size(input);
population = zeros(nPop, 3, class, features); % parameter: population size * 6 * total classes * total features
for loop=1:nPop
    a = zeros(class, features);
    b = repmat(2, class, features);
    c = zeros(class, features);
            for k = 1:class
                for i = 1:features % looping for all features
                    % premise parameter: a
                    
                    Ri=max(input(:, i))-min(input(:, i));
                    r=sum(Hi' == k);
                    aTemp = (Ri)/(2*r-2);
                    aLower = aTemp*0.5;
                    aUpper = aTemp*1.5;
                    a(k, i) = (aUpper-aLower).*rand()+aLower;

                    %premise parameter: c
                    dcc = (2.1-1.9).*rand()+1.9;
                    cLower = center(k,features)-dcc/2;
                    cUpper = center(k,features)+dcc/2;
                    c(k,i) = (cUpper-cLower).*rand()+cLower;
                end
            end
            population(loop, 1, :, :) = a;
            population(loop, 2, :, :) = b;
            population(loop, 3, :, :) = c;
end
end

File: getPopulation.py
import numpy as np
import skfuzzy as fuzz

def get_population(n_class, input_data, n_pop):
    """
    Generate initial population for optimization
    
    Parameters:
    -----------
    n_class : int
        Number of classes
    input_data : numpy.ndarray
        Input data, samples x features
    n_pop : int
        Population size
        
    Returns:
    --------
    population : numpy.ndarray
        Initial population with shape (n_pop, 3, n_class, features)
    """
    center, u, _, _, _, _, _ = fuzz.cluster.cmeans(
        input_data.T, 3, 2, error=1e-5, maxiter=100
    )  # center = center cluster, u = membership level
    
    hi = np.argmax(u, axis=0)  # the class corresponding to the max value
    samples, features = input_data.shape
    population = np.zeros((n_pop, 3, n_class, features))
    
    for loop in range(n_pop):
        a = np.zeros((n_class, features))
        b = np.full((n_class, features), 2)
        c = np.zeros((n_class, features))
        
        for k in range(n_class):
            for i in range(features):
                # premise parameter: a
                Ri = np.max(input_data[:, i]) - np.min(input_data[:, i])
                r = np.sum(hi == k)
                a_temp = Ri / (2 * r - 2) if r > 1 else Ri / 2
                a_lower = a_temp * 0.5
                a_upper = a_temp * 1.5
                a[k, i] = np.random.uniform(a_lower, a_upper)
                
                # premise parameter: c
                dcc = np.random.uniform(1.9, 2.1)
                c_lower = center[k, features-1] - dcc / 2
                c_upper = center[k, features-1] + dcc / 2
                c[k, i] = np.random.uniform(c_lower, c_upper)
        
        population[loop, 0, :, :] = a
        population[loop, 1, :, :] = b
        population[loop, 2, :, :] = c
    
    return population
